### ISSUE TITLE
Scheduled Master Pull Request 

### ISSUE THREAD

**Issue by bodinsamuel** [renovatebot/renovate#4408]
I'm sorry in if this is a duplicate. I tried searching for what I want but no luck.

**What would you like Renovate to be able to do?**
In our team we have a different process than most (I think):
- Renovate is ran on the weekend
- `baseBranches` is `["chore/renovateBaseBranch"]`
- Each monday, we open a PR with all the dependencies update made in the weekend
- We squash merge, to only have one commit `deps: update [week-number]`

This is working great to have a clear commit log, and be able to check if the whole system works with all the updates at the same time (meaning not loosing time checking out N branches, relaunching our UI and tests...).
There is a few issue with this: 
- `baseBranches` is not keep up to date with master automatically by renovate, so deps are tested against an old baseBranch until we do the PR and merge master in it.
- if we delete the branch after the PR is merge, renovate will complain (rightfully) that the baseBranch does not exists anymore. But if we don't delete it, recreating a PR next monday will put back every commit from the previous monday. And we can not create an empty PR in github.
 Solution was to push-force to reset the branch but not handy at all.

**Describe the solution you'd like**
A `masterPR` property that would:
- create the `baseBranch` automatically at start of process (require the use of `schedule`) `chore/renovate-masterpr-<date>` (or specified format)
- merge with master regularly during the process (or specified branch)
- automatic creation of a master Pull Request when the process start, e.g: `deps: update <date>` (or specified format)
- once the PR is merged, do not complain about the branch being deleted.
- rince and repeat on every `schedule`

```javascript
{
  baseBranches: ["master"],
  masterPR: {
    branchName: "chore/renovate-masterpr-{{date}}",
    prTitle: "deps: update {{date}}",
    updateWithBaseBranche: true,
    deleteOnMerge: true,
    manualMerge: true,
  }
}
```

**Describe alternatives you've considered**
A homemade script ahah

**Comment by rarkins** [comment]
There seems to be some overlap in ideas between this and #4404. I hadn't thought about the idea of using a Renovate "baseBranch" as a target for the PRs that roll up.

**Comment by bodinsamuel** [comment]
Indeed, keywords are different, but idea is somewhat similar ☺️ 
A `masterPR` echo your masterIssue feature and cover the **monorepo** problem while giving the flexibility to still tests failing dependencies individually.
The tricky part would be to tied it up to `schedule`, that would not play well with `every day` update. 

I haven't seen the `groupName` feature, it can be an alternative solution, e.g: 
```javascript
"packageRules": [
    {
      "packagePatterns": "*",
      "groupName": "everything"
    }
  ]
```
That would not help solve most of the issues we have, but considerably reduce noise

**Comment by bodinsamuel** [comment]
@rarkins did you decide on something? (wether this or #4404) ☺️

**Comment by rarkins** [comment]
Not yet. I got held up with a few "edge conditions" that are really tough. Mainly around what happens when new versions appear that haven't been tested yet. e.g. do you keep the old version in, or do you invalidate the master PR temporarily, etc.

**Comment by bodinsamuel** [comment]
Hi,
Just pinging to see if you had any traction on this? ☺️

**Comment by bodinsamuel** [comment]
Just want to see we have been successfully running this manually:
- a cron (in circle ci) that creates a brand new branch + PR on friday evening every week
- configured renovate to merge everything in this PR

This makes everything way simpler for us, all updates passing are simply merged on this PR, the rest not passing are just locked until resolution. We can test and merge only one single PR to master which reduce noise in the history, and we keep track of our weekly update easily (with `masterIssue`)

```typescript
// circleci.yml
  renovate_weekly_pr:
    triggers:
      - schedule:
          cron: '0 16 * * 5'
          filters:
            branches:
              only:
                - master

// weeklyPR.ts
// To use: 
//    GH_TOKEN=YOURTOKEN yarn ts-node weeklyPR.ts
// 
import { Octokit } from '@octokit/rest';

import { wait } from '@algolia/crawler-common/src/wait';

const GITHUB_TOKEN = process.env.GH_TOKEN;
const OWNER = 'algolia';
const REPO = 'test';
const BRANCH = 'chore/renovateBaseBranch';
const BRANCH_BASE = 'master';
const EMPTY_COMMIT_MSG = 'Automatic empty commit';

async function getRef(
  octokit: Octokit,
  branch: string
): Promise<string | false> {
  try {
    const ref = await octokit.git.getRef({
      owner: OWNER,
      repo: REPO,
      ref: `heads/${branch}`,
    });
    return ref.data.object.sha;
  } catch (err) {
    if (err.status !== 404) {
      throw err;
    }
  }
  return false;
}

async function createBranch(octokit: Octokit, sha: string): Promise<any> {
  const create = await octokit.git.createRef({
    owner: OWNER,
    repo: REPO,
    ref: `refs/heads/${BRANCH}`,
    sha,
  });
  return create;
}

async function deleteRef(octokit: Octokit): Promise<any> {
  console.log(`Deleting ref for ${BRANCH}`);
  const ref = await octokit.git.deleteRef({
    owner: OWNER,
    repo: REPO,
    ref: `heads/${BRANCH}`,
  });
  return ref;
}

async function updateRef(
  octokit: Octokit,
  sha: string
  // force: boolean = false
): Promise<any> {
  console.log(`Changing ref for ${BRANCH} to`, sha);
  const ref = await octokit.git.updateRef({
    owner: OWNER,
    repo: REPO,
    ref: `heads/${BRANCH}`,
    sha,
    // force,
  });
  return ref;
}

async function getCommit(octokit: Octokit, sha: string): Promise<any> {
  const commit = await octokit.git.getCommit({
    owner: OWNER,
    repo: REPO,
    // eslint-disable-next-line @typescript-eslint/camelcase
    commit_sha: sha,
  });
  return commit.data;
}

function isCommitAnEmptyCommit(commit: any): boolean {
  return commit.message.search(EMPTY_COMMIT_MSG) >= 0;
}

async function createEmptyCommit(
  octokit: Octokit,
  refCommit: any
): Promise<any> {
  console.log('Creating empty commit');
  const commit = await octokit.git.createCommit({
    owner: OWNER,
    repo: REPO,
    message: EMPTY_COMMIT_MSG,
    tree: refCommit.tree.sha,
    parents: [refCommit.sha],
  });
  return commit.data;
}

async function createPR(octokit: Octokit): Promise<any> {
  // Next monday
  const date = new Date();
  date.setDate(date.getDate() + 3);

  const title = `chore: dependencies ${date.toISOString().split('T')[0]}`;
  const { data } = await octokit.pulls.create({
    repo: REPO,
    owner: OWNER,
    title,
    body: `Weekly dependencies update.
Contributes to #5336
    `,
    head: BRANCH,
    base: BRANCH_BASE,
  });
  return data;
}

async function resetBranch(
  octokit: Octokit,
  refBase: string,
  exists: boolean
): Promise<void> {
  if (exists) {
    console.log('Deleting branch');
    await deleteRef(octokit);
    await wait(5000);
  }

  console.log('Creating branch');

  await createBranch(octokit, refBase);

  const commit = await getCommit(octokit, refBase);

  const empty = await createEmptyCommit(octokit, commit);
  await updateRef(octokit, empty.sha);
}

(async (): Promise<void> => {
  try {
    const octokit = new Octokit({
      auth: `token ${GITHUB_TOKEN}`,
    });

    const refBase = await getRef(octokit, BRANCH_BASE);
    const refTarget = await getRef(octokit, BRANCH);
    console.log(BRANCH_BASE, 'is at', refBase);
    console.log(BRANCH, 'is at', refTarget);

    if (!refBase) {
      console.error('no sha for base branch');
      return;
    }

    if (refTarget) {
      console.log('Branch exists');
      const commit = await getCommit(octokit, refTarget);

      if (isCommitAnEmptyCommit(commit)) {
        console.log('Empty commit exists');
        return;
      }
    }

    await resetBranch(octokit, refBase, Boolean(refTarget));

    console.log('Creating pull request');
    await createPR(octokit);
  } catch (e) {
    console.error(e);
  }
})();
```

<img width="902" alt="Screenshot 2020-07-26 at 18 50 39" src="https://user-images.githubusercontent.com/1637651/88484725-0330b800-cf71-11ea-8c67-02b9950268e7.png">

**Comment by rarkins** [comment]
@bodinsamuel that's really great, thanks for sharing. What are your thoughts on scenarios like the following?
 - If the combined PR gets out of date with the base branch it's to be merged into
 - If the combined PR gets in conflict with the base branch it's to be merged into
 - If the combined PR starts *failing* status checks due to a conflict in two upgraded dependencies (both of which passed their own tests, but together they fail and take down the entire PR)

In such cases do you think manual intervention is acceptable?

**Comment by bodinsamuel** [comment]
> If the combined PR gets out of date with the base branch it's to be merged into

Right now the cron simply erase the branch each friday so there is no "out of date" possible :D 
If it takes you a few days to merge the branch you can still manually merge master though.
If it takes you more than 6 days to merge, the branch is erased so renovate will have to do everything again (which could be fixed in my script but I didn't really bother).

> If the combined PR gets in conflict with the base branch it's to be merged into

manual is okay I guess, even though a "rebase" button could be cool

> If the combined PR starts failing status checks due to a conflict in two upgraded dependencies (both of which passed their own tests, but together they fail and take down the entire PR)

Manual intervention is completely fine, I would even say that almost what this feature imply as you explicitly request control over the final merge.

**Comment by HonkingGoose** [comment]
> There seems to be some overlap in ideas between this and #4404. I hadn't thought about the idea of using a Renovate "baseBranch" as a target for the PRs that roll up.

@rarkins Do you want to keep this issue, or close it? This issue was last updated Jul 29, 2020 so it might be outdated now you've started work on #4404.

**Comment by rarkins** [comment]
I think let's experiment with #4404 once it's landed and then decide if there's any other work to be done for use cases like these

**Comment by bodinsamuel** [comment]
For the one interested I created a Github actions to partially solve this issue. At Algolia we have been using this strategy on multiple repos and it works well 👌🏻 

https://github.com/bodinsamuel/renovate-automatic-branch
